qutebrowser/browser/mhtml.py

```diff
@@ -0,0 +1,534 @@
+            # cases). Using "ignore" lets us decode the file even if it's
+            # invalid UTF-8 data.
+            # The file written to the MHTML file won't be modified by this
+            # decoding, since there we're taking the original bytestream.
+            try:
+                css_string = item.fileobj.getvalue().decode('utf-8')
+            except UnicodeDecodeError:
+                log.downloads.warning("Invalid UTF-8 data in %s", url)
+                css_string = item.fileobj.getvalue().decode('utf-8', 'ignore')
+            import_urls = _get_css_imports(css_string)
+            for import_url in import_urls:
+                absolute_url = url.resolved(QUrl(import_url))
+                self.fetch_url(absolute_url)
+
+        encode = E_QUOPRI if mime.startswith('text/') else E_BASE64
+        # Our MHTML handler refuses non-ASCII headers. This will replace every
+        # non-ASCII char with '?'. This is probably okay, as official Content-
+        # Type headers contain ASCII only anyway. Anything else is madness.
+        mime = utils.force_encoding(mime, 'ascii')
+        self.writer.add_file(urlutils.encoded_url(url),
+                             item.fileobj.getvalue(), mime, encode)
+        item.fileobj.actual_close()
+        if self.pending_downloads:
+            return
+        self.finish_file()
+
+    def error(self, url, item, *_args):
+        """Callback when a download error occurred.
+
+        Args:
+            url: The orignal url of the asset as QUrl.
+            item: The DownloadItem given by the DownloadManager.
+        """
+        try:
+            self.pending_downloads.remove((url, item))
+        except KeyError:
+            # This might happen if .collect_zombies() calls .finished() and the
+            # error handler will be called after .collect_zombies
+            log.downloads.debug("Oops! Download already gone: %s", item)
+            return
+        item.fileobj.actual_close()
+        # Add a stub file, see comment in .fetch_url() for more information
+        self.writer.add_file(urlutils.encoded_url(url), b'')
+        if self.pending_downloads:
+            return
+        self.finish_file()
+
+    def finish_file(self):
+        """Save the file to the filename given in __init__."""
+        if self._finished:
+            log.downloads.debug("finish_file called twice, ignored!")
+            return
+        self._finished = True
+        log.downloads.debug("All assets downloaded, ready to finish off!")
+        try:
+            file_output = open(self.dest, 'wb')
+        except OSError as error:
+            message.error(self._win_id,
+                          "Could not save file: {}".format(error))
+            return
+        else:
+            # Yes, ugly nesting of try-blocks, but necessary because there's
+            # no way to let the try block only affect the with-statement-
+            # entry-code and not the with-statement-body.
```

Comment:
`OSError` can also occur while writing, e.g. because the disk is full. So I think this would be better?

```python
try:
    with open(self.dest, w'b') as file_output:
        self.writer.write_to(file_output)
except OSError as error:
    message.error(self._win_id, "Could not save file: {}".format(error))
    return
...
```